JBPM-5649: Stunner - Service Tasks

romartin opened the pull request:
Hey @hasys @jomarko 

This is the stuff that provides support for Service Tasks in Stunner. It does also provide other fixes and refactors that were already well known, this way I'm trying to take profit from the last commits before releasing and having more code to maintain in other branches... so sorry about this big PR!! Anyway hope  i'll look good for you :+1: 

See the ticket for service tasks here -> https://issues.jboss.org/browse/JBPM-5649

**IMPORTANT**
- Merge with https://github.com/kiegroup/jbpm-wb/pull/996
- Merge with https://github.com/kiegroup/kie-wb-distributions/pull/707

This commit does:
- **Support for Work Item Definitions** / Service Tasks in Stunner (client & backend)
- **Palette refactorings** - Removed old unused models and removed all duplicity that was present for DMN, so dropped the palette implementation (copy/pasted) and do the delegation to the default Stunner one
- **Removed the built-in backend services** that Stunner was using by default on the workbench, and just use them for BPMN and DMN standalone showcases. It's due that on the workbench environment, the workbench itself provide that services, so it was existing some (wrong) duplicity, did not hurt but was making noise and consuming some resources. On the code side it's mostly about refactoring some backend classes to the `stunner-backend-common` module, and just keeping the `stunner-backend` the ones used for the standalone showcases (which do not depend on kie/guvnor services). This way I was able to remove the dependency to`stunner-backend` from the kie/drools workbenches.
- **SVG generation** support for Images
- Some **cleans** and fixes on the BPMN/DMN constants
- BPMN Error reference property - Label renamed to just "Error" (requested by Kris)
- BPMN Property labels renamig -> "message/signal ref" to just "message/signal" (requested by Kris)

There a few still missing tasks in order to synchronize the actual job done by both designers, so created this new ticket which follows up on this task -> https://issues.jboss.org/browse/JBPM-7037

**ABOUT TESTING**
- So as described on last Stunner's meeting, it means that in order to see and use the service tasks (for testing this) in stunner, you have to **previously open or create another process using the actual jbpm-desginer**, which will do the asset deployments for the service tasks. Once just any process opened with the old designer, you can go to Stunner and use the service tasks This limitation will be fixed as with the [JBPM-7037](https://issues.jboss.org/browse/JBPM-7037)
- I recommend using the `Mortgages` example project, which contains several service task types.

Thanks!!

romartin commented:
@hasys @jomarko @manstis This removes the DMN palette implementation, it was mostly the same one as in Stunner ... so I just fixed the Stunner one to suport it. I guess it's everything as before, anyway let you know that @wmedvede is actually working on this again, not sure if further changes on DMN will be required, I guess @wmedvede has more information if you need so.

@wmedvede Can you please start working on the palette on top of this please?? This will be not rebased yet until merging, I'll be adding commits on top meanwhile the review is in progress. Sorry for the inconveniences! 

Thanks guys

romartin commented:
Jenkins execute full downstream build

pefernan commented:
@romartin I see there's a fail on jenkins because of a forms related error... I'll try to fix it.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/shape/view/handler/ServiceTaskShapeViewHandler.java:
```diff
@@ -0,0 +1,75 @@
+                .getIconData();
+        final String iconData = null != itemIconData ? itemIconData : WorkItemDefinitionClientRegistry.DEFAULT_ICON_DATA;
+        // Obtain the image element from the svg shape view.
+        final SVGPrimitive svgPrimitive = SVGViewUtils.getPrimitive(view, WID_ICON_ID).get();
```
Yes, but in that case it's fine. The instance for the expected id MUST be present on the generated view from the SVG file, it's not something that we can decide at runtime. So in a case somebody changes the SVG but not this, I prefer having an NPE rather than no exception, at least to realize something is wrong.
So in this case, IMO this call to get is safe anyway.

romartin commented on kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/components/palette/DefaultPaletteUtils.java:
```diff
@@ -0,0 +1,43 @@
+                .flatMap(DefaultPaletteUtils::flattern)
+                .filter(item -> itemId.equals(item.getId()))
+                .findFirst()
+                .get()
```
Same as before. IMO this get is safe anyway. So better having an NPE rather than trying to handle here when the item is not found, this should not happen, so if it does, it's an error on the palette code somewhere, so let them throw the NPE

romartin commented:
Hey @hasys 
Thanks for the initial review. I've added commits - merged mostly everything from https://github.com/romartin/kie-wb-common/pull/7 and applied your comments.
Still looking into the issue when closing the editor, a popup with an an exception message appears for me as well. Anyway you can move forward with your tests on this, will provide the fix asap.
Thanks!

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey @hasys 
Added another commit that fixes the NPE when closing the diagram, and includes the unit tests for this as well. Thx!

romartin commented:
Jenkins execute full downstream build

romartin commented:
@hasys BTW I've identified a bug related to service tasks and morphing - actually in the morphing menu for a task appears a "service task", but it should not. As there are a few already known issues with morphing, let you know that I've added this new case on the existing [JBPM-6829](https://issues.jboss.org/browse/JBPM-6829), and we'll address the different issues about morphing at once during the following weeks. Is that good for you?
Thanks

romartin commented:
Hey @hasys 
Another commit added - I just realized that the palette for CM was wrong, it was showing a different structure. Just fixed this as well. This does not affect other palettes like BPMN / DMN anyway. Tested and all paletes are the expected ones right now.

So I think at this point, rather than missing the service tasks deployments ([JBPM-7037](https://issues.jboss.org/browse/JBPM-7037)) and the issues about morphing ([JBPM-6829](https://issues.jboss.org/browse/JBPM-6829)), everything is ok. Lemme know plz. Thanks!

romartin commented:
Jenkins execute full downstream build

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/components/palette/DMNPaletteDefinitionBuilder.java:
```diff
@@ -0,0 +1,76 @@
+    @PostConstruct
+    public void init() {
+        paletteDefinitionBuilder
+                .itemFilter(defId -> true)
```
Looking on `AbstractPaletteDefinitionBuilder#initDefaults` I think this line can be omitted.

romartin replied:
yep - good point, done :+1:  
thx

pefernan commented:
@kkufova you should be able to test it but the problems in your git repo are still there.

hasys commented:
Hi @romartin,

thank you, next round of review (still not finished):
* Properties renames (message, signal, error) - looks good, verified
* Default palette node - works good, verified
* Palette with Service tasks (customs icons) - works with known issues in IE11, just one note:
  * Icons looks little bit grainy (see Picture 1.) is it possible to do it more smooth? Do we need some assistance from UX team?

Also, I put a comment with question to the Jira.

Picture 1.
![image](https://user-images.githubusercontent.com/1477262/37605507-9b8d3e68-2b93-11e8-8688-7a0a28cf3b93.png)

romartin commented:
Hey @hasys ,

Just rebased from master added a new commit. Good point about the icon for the service task, let's see what @srambach proposes too.

The icon is grainy (in your screenshot) because I was scaling it a little bit... now I've changed it to do not scale, adding screenshot. The "problem" is that (for backward compatibility as well) the icons for service tasks are defined by the user, but our documentation (for the old designer) says that icons have to be `16x16px`. So using this size they're looking a bit smaller than the icons for other tasks (user, script, etc), this is why I was scaling a bit the icon to see how it looks. But finally I think that is better not scaling the icons, they look smaller but at least they're looking good. Not sure if that's good for you... WDYT?

![screenshot from 2018-03-20 00-48-51](https://user-images.githubusercontent.com/4602417/37629239-fa8d1af6-2bdd-11e8-846e-4b51fc17dc78.png)

romartin commented:
@hasys Anyway rather than the concrete icon details, this is ready again for being tested. Thanks!

romartin commented:
Hey @hasys 
Missing your review again here when you have a chance please. Both @evacchi  and @wmedvede have tasks that depend on this... so would like to merge asap. Thx!

hasys commented:
Hi @romartin,

sorry for the delay. Works like a charm to me (parsing WID, display on canvas/properties, runtime).
Just one minor issues were reported: [JBPM-7072](https://issues.jboss.org/browse/JBPM-7072), but I guess we can let this PR go and fix this issue as separate PR.

Also, Feature Request is reported for the future as discussed in Jira issue: [JBPM-7073](https://issues.jboss.org/browse/JBPM-7073)
